crates/red_knot_python_semantic/src/types/class.rs

```diff
@@ -802,38 +806,193 @@ impl<'db> ClassLiteralType<'db> {
-                    );
-                    return Symbol::bound(Type::Callable(CallableType::new(db, signature))).into();
+                    Some(KnownClass::Bool.to_instance(db)),
+                );
+
+                return Some(Type::Callable(CallableType::new(db, signature)));
+            }
+        }
+
+        None
+    }
+
+    /// Returns a list of all annotated attributes defined in this class, or any of its superclasses.
+    ///
+    /// See [`ClassLiteralType::own_dataclass_fields`] for more details.
+    fn dataclass_fields(
+        self,
+        db: &'db dyn Db,
+        specialization: Option<Specialization<'db>>,
+    ) -> FxOrderMap<Name, (Type<'db>, Option<Type<'db>>)> {
+        let dataclasses_in_mro: Vec<_> = self
+            .iter_mro(db, specialization)
+            .filter_map(|superclass| {
+                if let Some(class) = superclass.into_class() {
+                    let class_literal = class.class_literal(db).0;
+                    if class_literal.dataclass_metadata(db).is_some() {
+                        Some(class_literal)
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
+            })
+            // We need to collect into a `Vec` here because we iterate the MRO in reverse order
+            .collect();
+
+        dataclasses_in_mro
+            .into_iter()
+            .rev()
+            .flat_map(|class| class.own_dataclass_fields(db))
+            // We collect into a FxOrderMap here to deduplicate attributes
+            .collect()
+    }
+
+    /// Returns a list of all annotated attributes defined in the body of this class. This is similar
+    /// to the `__annotations__` attribute at runtime, but also contains default values.
+    ///
+    /// For a class body like
+    /// ```py
+    /// @dataclass
+    /// class C:
+    ///     x: int
+    ///     y: str = "a"
+    /// ```
+    /// we return a map `{"x": (int, None), "y": (str, Some(Literal["a"]))}`.
+    fn own_dataclass_fields(
+        self,
+        db: &'db dyn Db,
+    ) -> FxOrderMap<Name, (Type<'db>, Option<Type<'db>>)> {
+        let mut attributes = FxOrderMap::default();
+
+        let class_body_scope = self.body_scope(db);
+        let table = symbol_table(db, class_body_scope);
+
+        let use_def = use_def_map(db, class_body_scope);
+        for (symbol_id, declarations) in use_def.all_public_declarations() {
+            if !declarations
+                .clone()
+                .all(|DeclarationWithConstraint { declaration, .. }| {
+                    declaration.is_some_and(|declaration| {
+                        matches!(
+                            declaration.kind(db),
+                            DefinitionKind::AnnotatedAssignment(..)
+                        )
+                    })
+                })
+            {
+                continue;
+            }
```

Comment:
This is a bit of a hack to avoid things like function definitions and nested class definitions showing up as dataclass fields. The filtering by `AnnotatedAssignment` is correct, I think, but we do not correctly handle weird things like
```py
class C:
    if flag():
        def attr(): ...
    else:
        attr: int = 1
```

Another option to solve this might be to pass down some kind of definition-kind-filter to `symbol_from_declarations`? Thoughts?

Reply:
A filter for `symbol_from_declarations` seems reasonable -- or it could even be a filter in the use-def map fetching?

The even trickier part about supporting something like this is that I think it would mean we'd have to generate a union of `__init__` methods with different signatures? It seems closely-related to the possibly-unbound handling you mention below, in that sense.

I definitely think this can be a TODO for now, I don't think any other type checker supports it.

Reply:
I added a comment for now.